### PR TITLE
LPS-113302 Apply classic ckeditor implementation in Message Boards, part 1/2

### PR DESCRIPTION
**Pull request by markocikos**
This is a followup on https://github.com/liferay-frontend/liferay-portal/pull/107

Changes:
- Removed dependency on `resources.jsp`. To achieve this:
  - [the ugly one] Added manual `ckeditor` script loading, to ensure `CKEDITOR` global object will be available.
  - Added cleanup of `CKEDITOR` global object when navigating away from a page with editor. I did not migrate manual tracking of instances count, `addInstance`, `removeInstance`, `ckEditorInstances` etc. We can see that out-of the-box in `CKEDITOR.instances` array. I am not sure if there is some point to it that I'm not seeing. It does not help with race condition on cleanup that I mentioned in https://github.com/liferay-frontend/liferay-portal/pull/107.
- Added debouncing on resize.
- `ckeditor` remains as component property.

Now, why the ugly loading change? Steps of the problem:
1. On page-1 load (reload), editor works normally.
1. Navigate to page-2 with ckeditor.
1. Navigate to page-1. Navigating away from page-2 deletes `CKEDITOR` object.

If we are using `resources.jsp`, this is not a problem, because we [reload `ckeditor.js` manually](https://github.com/liferay/liferay-portal/blob/e67026449a8a6dae28fc7a7da27e3ff6a3f780b6/modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/resources.jsp#L38).

In the new component, we rely on `ckeditor4-react` for loading. [This is the code](https://github.com/ckeditor/ckeditor4-react/blob/master/src/getEditorNamespace.js), and looking at it, it should work fine. If `CKEDITOR` is not defined, we run `loadScript` and re-initialize it. The Promise then resolves with sparkling `CKEDITOR` object.

But, what happens, promise returns with `CKEDITOR` = undefined and editor breaks on the next line, when we [invoke `onBeforeLoad`](https://github.com/ckeditor/ckeditor4-react/blob/470e9ba60aac1df5ae9f54016b906b06d496eade/src/ckeditor.jsx#L61).

This is quite an awesome bug, see the console below:
screen

It looks like window is not treated like a normal object:
```js
undefined in window // true
undefined in {} // false
```
JavaScript is awesome?  ¯\\_(ツ)_/¯

### CODE REVIEW

**Comment by markocikos** [comment]
ci:test:sf

**Comment by markocikos** [comment]
ci:test:relevant

**Comment by wincent** [comment]
> It looks like window is not treated like a normal object:
> 
> ```js
> undefined in window // true
> undefined in {} // false
> ```
> 
> JavaScript is awesome? ¯\_(ツ)_/¯

I don't think there's anything special about `window`. `undefined` is a property of it, and when you do `in` the left-hand-side gets stringified:

```js
window.hasOwnProperty('undefined') // true
```

This is also why people do tricks like write `void 0` instead of `undefined`, because some malicious actor could have done:

```js
window.undefined = 'haha';
```

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/Editor.js [hunk 1]
```diff
@@ -13,15 +13,54 @@
...
+
+	useEffect(() => {
+		let script;
+
+		const destroyGlobalCkEditor = () => {
+			if (
+				window.CKEDITOR &&
+				Object.keys(window.CKEDITOR.instances).length === 0
+			) {
+				window.CKEDITOR = undefined;
```
Comment (wincent) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/Editor.js:32]: Instead of assigment, you could actually `delete` it:

```js
window.CKEDITOR = undefined
'CKEDITOR' in window // true
delete window.CKEDITOR
'CKEDITOR' in window // false
```

Which may explain why the `in` check wasn't working before.

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/Editor.js [hunk 2]
```diff
@@ -13,15 +13,54 @@
...
+
+		if (!initialized) {
+			script = document.createElement('script');
+
+			script.setAttribute('data-senna-track', 'temporary');
+			script.src = CKEDITOR_PATH;
+
+			script.onload = () => {
+				setInitialized(true);
+			};
```
Comment (wincent) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/Editor.js:46]: As this is async, probably want to guard the `setInitialized()` call in an `isMounted()` check from the `useIsMounted` hook.

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/Editor.js [hunk 3]
```diff
@@ -13,15 +13,54 @@
...
+const CKEDITOR_PATH = `${CKEDITOR_BASEPATH}ckeditor.js`;
 
 const Editor = React.forwardRef((props, ref) => {
-	return <CKEditor ref={ref} {...props} />;
+	const [initialized, setInitialized] = useState(!!window.CKEDITOR);
+
+	useEffect(() => {
+		let script;
+
+		const destroyGlobalCkEditor = () => {
```
Comment (wincent) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/Editor.js:27]: As this function only depends on global state you could hoist it up to the top-level of the module to avoid recreating it whenever the effect runs.

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/Editor.js [hunk 4]
```diff
@@ -13,15 +13,54 @@
...
+			if (
+				window.CKEDITOR &&
+				Object.keys(window.CKEDITOR.instances).length === 0
+			) {
+				window.CKEDITOR = undefined;
+
+				Liferay.detach('beforeScreenFlip', destroyGlobalCkEditor);
```
Comment (wincent) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/Editor.js:34]: You can probably avoid an explicit `Liferay.detach()` by changing the `Liferay.on()` below to `Liferay.once()`. One less thing to worry about (for example, you don't have to worry about the stability of the `destroyGlobalCKEditor` function, which must be maintained in order for `detach()` to actually work).

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/Editor.js (hunk 1 above)
Comment (markocikos) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/Editor.js:32]: That's it! ~~Javascript~~ Greg is awesome! A simple thing, when you stop to think :slightly_smiling_face:. This means we don't need the ugly loading, just need to make this change in a few places.

Comment (jbalsas) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/Editor.js:32]: Yup! 👍

**Comment by jbalsas** [comment]
I assume @markocikos is now _prettyfying_ this back, so going to label as `need-update` 😉

**Comment by markocikos** [comment]
Followup in https://github.com/liferay-frontend/liferay-portal/pull/127

**Comment by jbalsas** [comment]
Hey @markocikos, just out of curiosity... why are you choosing to create new PRs instead of simply (force) pushing here and ask for comments?

**Comment by markocikos** [comment]
> Hey @markocikos, just out of curiosity... why are you choosing to create new PRs instead of simply (force) pushing here and ask for comments?

@jbalsas 
1. To preserve full context of the inline comments. For an outdated comment we only have a couple of lines.
1. I do it when there a big difference if code. Then, it makes sense to me to group discussion with related code. If a thread contains several discussions, where all but the last one makes no sense with the code in commits, it's poorly readable down the line.

What's a 'big' difference is a subjective call. Sometimes I do push additional commit, or force push, specially if nobody commented inline yet.

**Comment by jbalsas** [comment]
Sounds good, I was mostly just curious... thinking about how can we tweak the process if necessary...

**Comment by wincent** [comment]
It's a trade-off: preserving context at the cost of spreading it around, and I'm glad to hear that @markocikos is being so intentional about it.

> thinking about how can we tweak the process if necessary

[Maybe](https://www.phacility.com/phabricator/differential/). [One](https://www.gerritcodereview.com/). [Day](https://reviewable.io/). (Disclaimer: I haven't tried that last one, but I have worked with the first two.)
